grails-app/services/org/pih/warehouse/product/ProductService.groovy

```diff
@@ -390,9 +401,50 @@ class ProductService {
             if (sortColumn) order(sortColumn, sortOrder)
         }
 
-        return results
+		if (catalogsInput && !tagsInput) {
+			results = getProductsByCatalogs(catalogsInput, category, params.includeCategoryChildren)
+		} else {
+			results += getProductsByCatalogs(catalogsInput, category, params.includeCategoryChildren)
+		}
+
+        return results.unique()
     }
 
+	/**
+	 * Get all products in the given catalog
+	 *
+	 * @param inputCatalogs
+	 * @param category
+	 * @param includeCategoryChildren
+	 * @return
+	 */
+	def getProductsByCatalogs(List<ProductCatalog> inputCatalogs, Category category, includeCategoryChildren) {
```

Comment:
Pulling all of this data into memory to implement your own database query is inefficient. This should either be integrated into another query or written as its own query. 

This is what the query might look like if I used your code as requirements. 
```
def getProductsByCatalogs(List<ProductCatalog> productCatalogs, Category category, includeCategoryChildren)

    def categories = includeCategoryChildren ? 
        category.someMethodThatFetchesThisCategoryAndItsChildren() : [category]

    def products = ProductCatalogItem.createCriteria().list { 
        projections { 
            property("product")
        }
        product {
            "in"("category", categories)
        }
        "in"("productCatalog", productCatalogs)        
    }
    
    return products
}
```

However, taking a step back I feel like this would probably be bad for the overall feature we're implementing (search by tag, category, catalog, etc) because we are combining results from multiple queries in a way that does not support pagination, not to mention more complex composition (i.e. ANDing vs ORing the results).